File: python/71-80/Word Search.py
class Solution:
    def search(self,i,j,base,word, board):
        base[i][j] =1
        if len(word)<1:
            return True
        up,down,left,right=False,False,False,False
        if i!=0 and base[i-1][j]==0 and word[0]==board[i-1][j]:
            #temp = copy.deepcopy(base)
            up = self.search(i-1,j,base[:],word[1:],board)
            if up:
                return True
        if i!=len(base)-1 and base[i+1][j]==0 and word[0]==board[i+1][j]:
            down = self.search(i+1,j,base[:],word[1:],board)
            if down :
                return True
        if j!=0 and base[i][j-1]==0 and word[0]==board[i][j-1]:
            left = self.search(i,j-1,base[:],word[1:],board)
            if left:
                return True
        if j!=len(base[0])-1 and base[i][j+1]==0 and word[0]==board[i][j+1]:
            right = self.search(i,j+1,base[:],word[1:],board)
            if right :
                return True
        base[i][j]=0   #shallow copy  so need to change back
        return False
    # @param board, a list of lists of 1 length string
    # @param word, a string
    # @return a boolean
    def exist(self, board, word):
        if len(board)<1 or len(board[0])<1:
            return False
        if len(word)<1:
            return True
        row,col = len(board),len(board[0])
        base = [[0 for j in range(0,col)] for i in range(0,row)]
        for i in range(0,row):
            for j in range(0,col):
                if board[i][j]==word[0]:
                    if self.search(i,j,base[:],word[1:],board):
                        return True
        return False
                        

File: Cpp/71-80/word Search.cpp
/**
 question :https://leetcode.com/problems/word-search/
 finish time : 35min
 error before ac: 0
 time complexity: O(row*col*len(word))
 space complexity: O(1)
*/

class Solution {
public:
    bool dfs(vector<vector<bool> >&base, vector<vector<char>>& board, string word,int ind,int x, int y){
      if(ind==word.size())
        return true;
      base[x][y]=true;
      int row = board.size();
      int col = board[0].size();
      if(x>0&&!base[x-1][y]&&board[x-1][y]==word[ind])
        if(dfs(base,board,word,ind+1,x-1,y))
          return true;
      if(x<row-1&&!base[x+1][y]&&board[x+1][y]==word[ind])
        if(dfs(base,board,word,ind+1,x+1,y))
          return true;
      if(y>0&&!base[x][y-1]&&board[x][y-1]==word[ind])
        if(dfs(base,board,word,ind+1,x,y-1))
          return true;
      if(y<col-1&&!base[x][y+1]&&board[x][y+1]==word[ind])
        if(dfs(base,board,word,ind+1,x,y+1))
          return true;
      base[x][y]=false;
      return false;
    }
    bool exist(vector<vector<char>>& board, string word) {
      if(board.size()==0|| board[0].size()==0||word.size()==0)      
        return false;
      int row,col;
      row = board.size();
      col = board[0].size();
      if(row*col < word.size())
        return false;
      //iteration the 2d board
      vector<vector<bool> > base(row,vector<bool>(col,false));
      for(int i=0; i<row; i++){
        for(int j =0; j<col; j++){
          if(board[i][j]==word[0])
            if(dfs(base,board,word,1,i,j))
                return true;
        }
      }
      return false;
    }
};
